IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/WebElement.cs

```diff
@@ -0,0 +1,253 @@
+                    {
+                        WrappedElement = _Driver.FindElement(By);
+                    }
+                    string result = null;
+                    switch (action)
+                    {
+                        case "click":
+                            WrappedElement.Click();
+                            return null;
+                        case "submit":
+                            WrappedElement.Submit();
+                            return null;
+                        case "sendkeys":
+                            WrappedElement.SendKeys(text);
+                            return null;
+                        case "getattribute":
+                            result = WrappedElement.GetAttribute(text);
+                            return result;
+                        case "clear":
+                            WrappedElement.Clear();
+                            return null;
+                        case "getcss":
+                            result = WrappedElement.GetCssValue(text);
+                            return result;
+                        case "getproperty":
+                            return WrappedElement.GetProperty(text);
+                        case "text":
+                            result = WrappedElement.Text;
+                            return result;
+                        case "enabled":
+                            result = WrappedElement.Enabled.ToString();
+                            return result;
+                        case "displayed":
+                            result = WrappedElement.Displayed.ToString();
+                            return result;
+                        case "selected":
+                            result = WrappedElement.Selected.ToString();
+                            return result;
+                        default:
+                            throw new InvalidOperationException(
+                                "Unknown type of Selenium action passed to WebElement.RetryAction");
+                    }
+                }
+                catch (ElementNotVisibleException e)
+                {
+                    ex = e;
+                }
+                catch (StaleElementReferenceException e)
+                {
+                    ex = e;
+                }
+                catch (InvalidElementStateException e)
+                {
+                    ex = e;
+                }
+                catch (InvalidOperationException e)
+                {
+                    ex = e;
+                }
+                catch (NoSuchElementException e)
+                {
+                    ex = e;
+                }
+                catch (InvalidCastException e)
+                {
+                    ex = e;
+                }
+                catch (WebDriverException e)
+                {
+                    ex = e;
+                }
+
+                // Re-find the element to make sure it's valid.
```

Comment:
Why re-find the element if the method is just going to throw?

Reply:
For cases where Selenium tried to pick up the element while it was loading. Basically, the use case of we were expecting the element to be there, but maybe it took a second longer to load than before, so we re-find it in the loop to give us a chance of picking it up after it's in the right state.

HOWEVER, when this loop was first written, we knew less about the element we were finding, so I'm thinking I might be able to offload the "re-find" to the start of the loop, and find it only if it's not in the correct state using a check on the Initialized property.

Reply:
But I am not sure that is what this code does. The only way to get this far in the method is for an exception to be thrown and caught above (meaning the `ex` variable will be set). It then attempts this FindElement, which may or may not succeed and set the `WrappedElement` property. But regardless of whether this second find succeeds it will always hin the `if (ex != null)` below, and since we know that `ex` is set, this method will throw.

Reply:
I'm on my phone, so it's super hard to verify, but the throw *should* be outside the while loop to prevent that. If it's not, I will change it as soon as I get home and be extra diligent in checking my curly braces. 🙂

Reply:
So you are correct, the `throw` is outside of the `while` loop. However that doesn't actually fix the issue. So let's say you get here and do a second `FindElement` call and it doesn't throw an exception. That means we either exit the while loop or loop again.
If we exit the while loop, my above comment holds, and the method throws. If we loop again, the `ex` variable gets cleared, and we end up replacing the value we set in `WrappedElement`.

Reply:
Oooh, okay, makes sense. I'll continue ahead with looking into moving the re-find logic to the start of the loop and only if the element is in a bad state.